nni/common/serializer.py

```diff
@@ -0,0 +1,369 @@
+import base64
+import functools
+import inspect
+from typing import Any, Union, Dict, Optional, List, TypeVar
+
+import json_tricks  # use json_tricks as serializer backend
+import cloudpickle  # use cloudpickle as backend for unserializable types and instances
+
+
+__all__ = ['trace', 'dump', 'load', 'SerializableObject']
+
+
+T = TypeVar('T')
+
+
+class SerializableObject:
+    """
+    Serializable object is a wrapper of existing python objects, that supports dump and load easily.
+    Stores a symbol ``s`` and a dict of arguments ``args``, and the object can be restored with ``s(**args)``.
+    """
+
+    def __init__(self, symbol: T, args: List[Any], kwargs: Dict[str, Any],
+                 _self_contained: bool = False):
+        # use dict to avoid conflicts with user's getattr and setattr
+        self.__dict__['_nni_symbol'] = symbol
+        self.__dict__['_nni_args'] = args
+        self.__dict__['_nni_kwargs'] = kwargs
+
+        self.__dict__['_nni_self_contained'] = _self_contained
+
+        if _self_contained:
+            # this is for internal usage only.
+            # kwargs is used to init the full object in the same object as this one, for simpler implementation.
+            super().__init__(*self._recursive_init(args), **self._recursive_init(kwargs))
+
+    def get(self) -> Any:
+        """
+        Get the original object.
+        """
+        if self._get_nni_attr('self_contained'):
+            return self
+        if '_nni_cache' not in self.__dict__:
+            self.__dict__['_nni_cache'] = self._get_nni_attr('symbol')(
+                *self._recursive_init(self._get_nni_attr('args')),
+                **self._recursive_init(self._get_nni_attr('kwargs'))
+            )
+        return self.__dict__['_nni_cache']
+
+    def copy(self) -> Union[T, 'SerializableObject']:
+        """
+        Perform a shallow copy. Will throw away the self-contain property for classes (refer to implementation).
+        This is the one that should be used when you want to "mutate" a serializable object.
+        """
+        return SerializableObject(
+            self._get_nni_attr('symbol'),
+            self._get_nni_attr('args'),
+            self._get_nni_attr('kwargs')
+        )
+
+    def __json_encode__(self):
+        ret = {'__symbol__': _get_hybrid_cls_or_func_name(self._get_nni_attr('symbol'))}
```

Comment:
if falling back to cloudpickle, args and kwargs are useless and cannot be mutated, right?

Reply:
does cloudpickle work across different OSes?

Reply:
Yes.